File: typescript/src/ships/module-stat-labels.ts
/**
 * The bridge between journal Modifier Labels and the catalogue fields that hold their
 * base values — including the **unit and algebra** each label uses, which are not
 * always the catalogue's.
 *
 * Its own module so that code needing only the mapping — resolving a pre-engineered
 * variant's stats, say — does not pull in the blueprint and experimental-effect
 * catalogues alongside it.
 *
 * @internal
 */

import type { DamageDistribution, OutfittingModule } from './modules.js';

/** One journal modifier label and how it relates to the catalogue field behind it. */
export interface StatLabel {
    /**
     * The Modifier Label, e.g. `"FSDOptimalMass"` — a journal's own spelling, except for
     * `"BurstInterval"`, which blueprint recipes use for the stat a journal reports as
     * the resulting `"RateOfFire"`.
     */
    readonly label: string;
    /** The catalogue field holding the base value. */
    readonly field: keyof OutfittingModule;
    /**
     * Member within {@link DamageDistribution} when `field` is `damageDistribution`.
     * Absent for ordinary scalar module stats.
     */
    readonly damageType?: keyof DamageDistribution;
    /**
     * Journal value ÷ catalogue value. `1` for the stats both express the same way
     * (mass in tonnes, power in MW); `100` for the ones the journal reports as a
     * percentage while the catalogue stores a fraction (a 40% resistance is `40` in a
     * journal and `0.4` here).
     */
    readonly scale?: number;
    /**
     * For a stat that is really a **percentage of a multiplier**, the base the game
     * divides by before compounding — Frontier's `modmod`, `100` or `-100`.
     *
     * @remarks
     * These stats do not scale like an ordinary number. A bulkhead's `+80%` hull boost
     * engineered by a `+32%` blueprint becomes `137.6%`, not `105.6%`, because the
     * multiplier `1.8` is what gets multiplied by `1.32`. A resistance works the same
     * way on its *damage multiplier*, so the base is negative: a `-20%` kinetic
     * resistance (a ×1.2 damage multiplier) with a `+5%` blueprint becomes `-14%`
     * (×1.14). Both are folded by {@link computeModifiers}, whatever method the
     * blueprint declares.
     */
    readonly multiplierBase?: number;
    /**
     * The value the game assumes when a module does not carry this stat at all — a
     * weapon with no `roundsPerShot` fires one round per shot, and one with no burst
     * fires a single shot per burst. A recipe can modify a defaulted stat (Double Shot
     * gives a plain multi-cannon a two-round burst), so the default belongs in the base
     * values {@link computeModifiers} folds.
     */
    readonly defaultBase?: number;
}

const percent = 100;

/**
 * Journal modifier label → catalogue field holding its base value.
 *
 * Not injective in either direction by design: a module carries one optimal mass, and
 * which label names it depends on the module's kind, so several labels share `optMass`.
 *
 * @internal
 */
export const STAT_LABELS: readonly StatLabel[] = [
    { label: 'Mass', field: 'mass' },
    { label: 'Integrity', field: 'integrity' },
    { label: 'PowerDraw', field: 'powerDraw' },
    { label: 'BootTime', field: 'bootTime' },

    // ── Core performance ────────────────────────────────────────────────────
    { label: 'FSDOptimalMass', field: 'optMass' },
    { label: 'EngineOptimalMass', field: 'optMass' },
    { label: 'EngineMinimumMass', field: 'minMass' },
    { label: 'MaximumMass', field: 'maxMass' },
    { label: 'ShieldGenOptimalMass', field: 'optMass' },
    { label: 'ShieldGenMinimumMass', field: 'minMass' },
    { label: 'ShieldGenMaximumMass', field: 'maxMass' },
    { label: 'EngineOptPerformance', field: 'optMultiplier', scale: percent },
    { label: 'EngineMinPerformance', field: 'minMultiplier', scale: percent },
    { label: 'EngineMaxPerformance', field: 'maxMultiplier', scale: percent },
    { label: 'ShieldGenStrength', field: 'optMultiplier', scale: percent },
    { label: 'ShieldGenMinStrength', field: 'minMultiplier', scale: percent },
    { label: 'ShieldGenMaxStrength', field: 'maxMultiplier', scale: percent },
    { label: 'EngineHeatRate', field: 'engineHeatRate' },
    { label: 'FSDHeatRate', field: 'fsdHeatRate' },
    { label: 'MaxFuelPerJump', field: 'maxFuel' },
    { label: 'PowerCapacity', field: 'powerCapacity' },
    { label: 'HeatEfficiency', field: 'heatEfficiency' },
    { label: 'EnginesCapacity', field: 'enginesCapacity' },
    { label: 'EnginesRecharge', field: 'enginesRecharge' },
    { label: 'SystemsCapacity', field: 'systemsCapacity' },
    { label: 'SystemsRecharge', field: 'systemsRecharge' },
    { label: 'WeaponsCapacity', field: 'weaponsCapacity' },
    { label: 'WeaponsRecharge', field: 'weaponsRecharge' },
    { label: 'RefuelRate', field: 'refuelRate' },
    { label: 'FuelCapacity', field: 'fuelCapacity' },
    { label: 'CargoCapacity', field: 'cargoCapacity' },
    { label: 'RegenRate', field: 'shieldRegenRate' },
    { label: 'BrokenRegenRate', field: 'shieldBrokenRegenRate' },
    // A shield generator's distributor draw is the systems-capacitor cost of one MJ per
    // second of regeneration, and the journal gives it its own label rather than reusing
    // a weapon's `DistributorDraw`. Same catalogue field, both spellings.
    { label: 'EnergyPerRegen', field: 'distributorDraw' },

    // ── Shield cell banks ───────────────────────────────────────────────────
    { label: 'ShieldBankReinforcement', field: 'shieldBankReinforcement' },
    { label: 'ShieldBankHeat', field: 'shieldBankHeat' },
    // A cell bank's heat is also the `thermalLoad` its record has always carried — the
    // same figure from the same upstream field — so a recipe that moves it moves both.
    { label: 'ShieldBankHeat', field: 'thermalLoad' },
    { label: 'ShieldBankSpinUp', field: 'shieldBankSpinUp' },
    { label: 'ShieldBankDuration', field: 'shieldBankDuration' },

    // ── Scanning, and the FSD interdictor ───────────────────────────────────
    // Utility scanners and sensor suites share one range field. The weapon range below
    // remains separate even though a journal can spell either kind `Range`.
    { label: 'ScannerRange', field: 'scannerRange' },
    { label: 'SensorTargetScanAngle', field: 'scanAngle' },
    // The utility scanners' scan cone is the same stat under the journal's other name.
    { label: 'MaxAngle', field: 'scanAngle' },
    { label: 'ScannerTimeToScan', field: 'scanTime' },
    // The blueprint recipe says `ProbeRadius`; a journal writes `DSS_PatchRadius` for
    // the same stat on the same module. Both resolve.
    { label: 'ProbeRadius', field: 'probeRadius' },
    { label: 'DSS_PatchRadius', field: 'probeRadius' },
    { label: 'FSDInterdictorFacingLimit', field: 'interdictorFacingLimit' },
    { label: 'FSDInterdictorRange', field: 'interdictorRange' },

    // ── Defence ─────────────────────────────────────────────────────────────
    {
        label: 'DefenceModifierShieldMultiplier',
        field: 'shieldBoost',
        scale: percent,
        multiplierBase: percent,
    },
    // On an armour module this scales the bulkhead's own hull boost, which the catalogue
    // carries. A hull reinforcement package carries none, and because this is a
    // percentage-of-a-multiplier stat that absence is itself a value — no hull boost is a
    // ×1 multiplier, 0% — so it resolves from zero and the recipe's bonus *is* the
    // result, which is how the game reads it and what a journal reports
    // (`OriginalValue: 0`). See `multiplierBase` and {@link computeModifiers}.
    {
        label: 'DefenceModifierHealthMultiplier',
        field: 'hullBoost',
        scale: percent,
        multiplierBase: percent,
    },
    { label: 'DefenceModifierHealthAddition', field: 'hullReinforcement' },
    { label: 'DefenceModifierShieldAddition', field: 'shieldAddition' },
    {
        label: 'KineticResistance',
        field: 'kineticResistance',
        scale: percent,
        multiplierBase: -percent,
    },
    // The journal spells thermal resistance "Thermic" — the one thermal stat that
    // does not read "Thermal".
    {
        label: 'ThermicResistance',
        field: 'thermalResistance',
        scale: percent,
        multiplierBase: -percent,
    },
    {
        label: 'ExplosiveResistance',
        field: 'explosiveResistance',
        scale: percent,
        multiplierBase: -percent,
    },
    {
        label: 'CausticResistance',
        field: 'causticResistance',
        scale: percent,
        multiplierBase: -percent,
    },
    { label: 'DamageProtection', field: 'moduleProtection', scale: percent },
    { label: 'ModuleDefenceAbsorption', field: 'moduleProtection', scale: percent },

    // ── Weapons ─────────────────────────────────────────────────────────────
    { label: 'Damage', field: 'damage' },
    { label: 'Rounds', field: 'roundsPerShot', defaultBase: 1 },
    { label: 'RoundsPerShot', field: 'roundsPerShot', defaultBase: 1 },
    { label: 'RateOfFire', field: 'rateOfFire' },
    { label: 'BurstInterval', field: 'burstInterval' },
    { label: 'BurstSize', field: 'burstRounds', defaultBase: 1 },
    { label: 'BurstRateOfFire', field: 'burstRateOfFire', defaultBase: 1 },
    { label: 'AmmoClipSize', field: 'clipSize' },
    { label: 'AmmoMaximum', field: 'ammoMaximum' },
    { label: 'ReloadTime', field: 'reloadTime' },
    { label: 'DistributorDraw', field: 'distributorDraw' },
    { label: 'ThermalLoad', field: 'thermalLoad' },
    { label: 'ArmourPenetration', field: 'armourPiercing' },
    // `Range` is a weapon's maximum range and the journal's alternate spelling of a
    // scanner's `ScannerRange`. A record resolves it to the field that family carries.
    { label: 'Range', field: 'maximumRange' },
    { label: 'Range', field: 'scannerRange' },
    { label: 'MaximumRange', field: 'maximumRange' },
    { label: 'FalloffRange', field: 'falloffRange' },
    // A journal spells the same stat `DamageFalloffRange`, where a blueprint recipe says
    // `FalloffRange`. Both resolve, as `ProbeRadius` / `DSS_PatchRadius` do.
    { label: 'DamageFalloffRange', field: 'falloffRange' },
    { label: 'ShotSpeed', field: 'shotSpeed' },
    // Damage-type shares are percentages in a journal and fractions in the catalogue.
    // They live one level below the module record, so `damageType` names the member the
    // otherwise-flat label mapping reads and writes.
    {
        label: '$Kinetic;',
        field: 'damageDistribution',
        damageType: 'kinetic',
        scale: percent,
        defaultBase: 0,
    },
    {
        label: '$Thermal;',
        field: 'damageDistribution',
        damageType: 'thermal',
        scale: percent,
        defaultBase: 0,
    },
    {
        label: '$Explosive;',
        field: 'damageDistribution',
        damageType: 'explosive',
        scale: percent,
        defaultBase: 0,
    },
    {
        label: '$Absolute;',
        field: 'damageDistribution',
        damageType: 'absolute',
        scale: percent,
        defaultBase: 0,
    },
    // A weapon that carries no jitter fires true, and Rapid Fire, its multi-cannon spelling
    // and Inertial Impact all give one — which a journal confirms, reporting
    // `OriginalValue: 0` for a missile rack whose record holds no such field.
    // {@link computeModifiers} already reaches the right result without this, because an
    // additive leg starts from zero on its own; the default is what makes the *base*
    // explicit, so a computed modifier states the same `OriginalValue` the game does
    // instead of omitting it.
    { label: 'Jitter', field: 'jitter', defaultBase: 0 },
];

/**
 * Every entry for a label, in declaration order. Usually one; `Range` and
 * `ShieldBankHeat` have two each, because the modules that carry those stats keep them
 * in different catalogue fields. The first entry is the label's own answer for
 * everything that does not depend on which module is being asked about.
 */
const BY_LABEL: ReadonlyMap<string, readonly StatLabel[]> = (() => {
    const entries = new Map<string, StatLabel[]>();
    for (const entry of STAT_LABELS) {
        const forLabel = entries.get(entry.label) ?? [];
        forLabel.push(entry);
        entries.set(entry.label, forLabel);
    }
    return entries;
})();

const LABELS_BY_FIELD: ReadonlyMap<keyof OutfittingModule, readonly string[]> = (() => {
    const labels = new Map<keyof OutfittingModule, string[]>();
    for (const entry of STAT_LABELS) {
        // Nested damage shares are read together by `damageDistributionFor`; exposing
        // them as scalar field labels would make `effectiveStat` return one share where
        // callers expect the whole record.
        if (entry.damageType !== undefined) continue;
        const fieldLabels = labels.get(entry.field) ?? [];
        fieldLabels.push(entry.label);
        labels.set(entry.field, fieldLabels);
    }
    return labels;
})();

/**
 * Convert catalogue stats to base values keyed by journal Modifier Label, **in the
 * journal's own units** — so the modifiers {@link computeModifiers} returns are
 * journal-shaped and can be written straight onto a fitted module.
 *
 * @internal
 */
export function baseStats(stats: OutfittingModule): Record<string, number> {
    const base: Record<string, number> = {};
    for (const { label, field, damageType, scale, defaultBase } of STAT_LABELS) {
        if (base[label] !== undefined) continue; // an earlier entry already answered
        const value =
            damageType === undefined ? stats[field] : stats.damageDistribution?.[damageType];
        if (typeof value === 'number') base[label] = value * (scale ?? 1);
        // A stat the module leaves out but the game still assumes a value for can be
        // engineered from that assumed value.
        else if (defaultBase !== undefined && isWeapon(stats)) base[label] = defaultBase;
    }
    return base;
}

/** Whether a record is a weapon — the only modules the defaulted stats apply to. */
function isWeapon(stats: OutfittingModule): boolean {
    return stats.damage !== undefined;
}

/**
 * The catalogue field a journal modifier label writes back to, or `null` when the
 * catalogue models no field for that label at all.
 *
 * @param label - The journal Modifier Label.
 * @param stats - The record the label is being resolved against, when there is one.
 * A label that maps to two fields — `Range` or `ShieldBankHeat` — answers with whichever
 * of them the record carries; without a record it answers with the first.
 *
 * @internal
 */
export function fieldForLabel(
    label: string,
    stats?: OutfittingModule,
): keyof OutfittingModule | null {
    const entries = BY_LABEL.get(label);
    if (!entries || entries.length === 0) return null;
    if (stats && entries.length > 1) {
        const carried = entries.find((entry) => typeof stats[entry.field] === 'number');
        if (carried) return carried.field;
    }
    return entries[0]!.field;
}

/**
 * The damage-distribution member a journal modifier label names, or `null` for an
 * ordinary scalar stat or an unknown label.
 *
 * @internal
 */
export function damageTypeForLabel(label: string): keyof DamageDistribution | null {
    return BY_LABEL.get(label)?.[0]?.damageType ?? null;
}

/**
 * Journal labels that name one damage-distribution member, in declaration order.
 *
 * @internal
 */
export function labelsForDamageType(type: keyof DamageDistribution): readonly string[] {
    return STAT_LABELS.filter((entry) => entry.damageType === type).map((entry) => entry.label);
}

/**
 * Journal value ÷ catalogue value for a label — `100` for the stats a journal reports
 * as a percentage, `1` for everything else (and for unknown labels).
 *
 * @internal
 */
export function scaleForLabel(label: string): number {
    return BY_LABEL.get(label)?.[0]?.scale ?? 1;
}

/**
 * The `modmod` base a label compounds through, or `null` for the ordinary stats that
 * scale directly. See {@link StatLabel.multiplierBase}.
 *
 * @internal
 */
export function multiplierBaseForLabel(label: string): number | null {
    return BY_LABEL.get(label)?.[0]?.multiplierBase ?? null;
}

/**
 * Every journal modifier label that writes to a catalogue field — the inverse of
 * {@link fieldForLabel}, and not single-valued: a shield generator's optimal mass is
 * `ShieldGenOptimalMass` while a drive's is `FSDOptimalMass`, and both land on
 * `optMass`.
 *
 * @internal
 */
export function labelsForField(field: keyof OutfittingModule): readonly string[] {
    return LABELS_BY_FIELD.get(field) ?? [];
}

File: dotnet/src/EliteDangerousAlmanac/Ships/Internal/ModuleStatLabels.cs
using System;
using System.Collections.Generic;

namespace EliteDangerousAlmanac.Ships.Internal;

/// <summary>
/// The bridge between journal modifier labels and the module stats that hold their base
/// values, including the unit and the algebra each label uses.
/// </summary>
/// <remarks>
/// <para>
/// The mapping is not one to one in either direction by design. A module carries one optimal
/// mass, and which label names it depends on the module's kind, so several labels share
/// <see cref="ModuleStat.OptMass"/>.
/// </para>
/// <para>
/// It is its own type so that code needing only the mapping does not pull in the blueprint
/// and experimental-effect catalogues alongside it.
/// </para>
/// </remarks>
internal static class ModuleStatLabels
{
    private const double Percent = 100;

    private static readonly StatLabel[] Labels =
    [
        new("Mass", ModuleStat.Mass),
        new("Integrity", ModuleStat.Integrity),
        new("PowerDraw", ModuleStat.PowerDraw),
        new("BootTime", ModuleStat.BootTime),

        // Core performance.
        new("FSDOptimalMass", ModuleStat.OptMass),
        new("EngineOptimalMass", ModuleStat.OptMass),
        new("EngineMinimumMass", ModuleStat.MinMass),
        new("MaximumMass", ModuleStat.MaxMass),
        new("ShieldGenOptimalMass", ModuleStat.OptMass),
        new("ShieldGenMinimumMass", ModuleStat.MinMass),
        new("ShieldGenMaximumMass", ModuleStat.MaxMass),
        new("EngineOptPerformance", ModuleStat.OptMultiplier) { Scale = Percent },
        new("EngineMinPerformance", ModuleStat.MinMultiplier) { Scale = Percent },
        new("EngineMaxPerformance", ModuleStat.MaxMultiplier) { Scale = Percent },
        new("ShieldGenStrength", ModuleStat.OptMultiplier) { Scale = Percent },
        new("ShieldGenMinStrength", ModuleStat.MinMultiplier) { Scale = Percent },
        new("ShieldGenMaxStrength", ModuleStat.MaxMultiplier) { Scale = Percent },
        new("EngineHeatRate", ModuleStat.EngineHeatRate),
        new("FSDHeatRate", ModuleStat.FsdHeatRate),
        new("MaxFuelPerJump", ModuleStat.MaxFuel),
        new("PowerCapacity", ModuleStat.PowerCapacity),
        new("HeatEfficiency", ModuleStat.HeatEfficiency),
        new("EnginesCapacity", ModuleStat.EnginesCapacity),
        new("EnginesRecharge", ModuleStat.EnginesRecharge),
        new("SystemsCapacity", ModuleStat.SystemsCapacity),
        new("SystemsRecharge", ModuleStat.SystemsRecharge),
        new("WeaponsCapacity", ModuleStat.WeaponsCapacity),
        new("WeaponsRecharge", ModuleStat.WeaponsRecharge),
        new("RefuelRate", ModuleStat.RefuelRate),

        // A journal spells a fuel scoop's rate FuelScoopRate, where the recipe that moves it
        // says RefuelRate. Both resolve, as FalloffRange and DamageFalloffRange do.
        new("FuelScoopRate", ModuleStat.RefuelRate),
        new("FuelCapacity", ModuleStat.FuelCapacity),
        new("CargoCapacity", ModuleStat.CargoCapacity),
        new("CabinCapacity", ModuleStat.CabinCapacity),
        new("RegenRate", ModuleStat.ShieldRegenRate),
        new("BrokenRegenRate", ModuleStat.ShieldBrokenRegenRate),

        // A shield generator's distributor draw is the systems-capacitor cost of one megajoule
        // per second of regeneration, and the journal gives it its own label rather than
        // reusing a weapon's DistributorDraw. Same stat, both spellings.
        new("EnergyPerRegen", ModuleStat.DistributorDraw),

        // Shield cell banks.
        new("ShieldBankReinforcement", ModuleStat.ShieldBankReinforcement),
        new("ShieldBankHeat", ModuleStat.ShieldBankHeat),

        // A cell bank's heat is also the thermal load its record carries, the same figure from
        // the same upstream field, so a recipe that moves one moves both.
        new("ShieldBankHeat", ModuleStat.ThermalLoad),
        new("ShieldBankSpinUp", ModuleStat.ShieldBankSpinUp),
        new("ShieldBankDuration", ModuleStat.ShieldBankDuration),

        // Scanning, and the frame shift drive interdictor.
        new("ScannerRange", ModuleStat.ScannerRange),
        new("SensorTargetScanAngle", ModuleStat.ScanAngle),

        // The utility scanners' scan cone is the same stat under the journal's other name.
        new("MaxAngle", ModuleStat.ScanAngle),
        new("ScannerTimeToScan", ModuleStat.ScanTime),

        // The blueprint recipe says ProbeRadius; a journal writes DSS_PatchRadius for the same
        // stat on the same module. Both resolve.
        new("ProbeRadius", ModuleStat.ProbeRadius),
        new("DSS_PatchRadius", ModuleStat.ProbeRadius),
        new("FSDInterdictorFacingLimit", ModuleStat.InterdictorFacingLimit),
        new("FSDInterdictorRange", ModuleStat.InterdictorRange),

        // Defence.
        new("DefenceModifierShieldMultiplier", ModuleStat.ShieldBoost)
        {
            Scale = Percent,
            MultiplierBase = Percent,
        },

        // On an armour module this scales the bulkhead's own hull boost. A hull reinforcement
        // package carries none, and because this is a percentage of a multiplier that absence
        // is itself a value: no hull boost is a multiplier of one, which is zero percent. The
        // recipe's bonus is therefore the whole result, which is how the game reads it and
        // what a journal reports.
        new("DefenceModifierHealthMultiplier", ModuleStat.HullBoost)
        {
            Scale = Percent,
            MultiplierBase = Percent,
        },
        new("DefenceModifierHealthAddition", ModuleStat.HullReinforcement),
        new("DefenceModifierShieldAddition", ModuleStat.ShieldAddition),
        new("GuardianModuleResistance", null) { CapabilityValue = "Active" },
        new("KineticResistance", ModuleStat.KineticResistance)
        {
            Scale = Percent,
            MultiplierBase = -Percent,
        },

        // The journal spells thermal resistance "Thermic", the one thermal stat that does not
        // read "Thermal".
        new("ThermicResistance", ModuleStat.ThermalResistance)
        {
            Scale = Percent,
            MultiplierBase = -Percent,
        },
        new("ExplosiveResistance", ModuleStat.ExplosiveResistance)
        {
            Scale = Percent,
            MultiplierBase = -Percent,
        },
        new("CausticResistance", ModuleStat.CausticResistance)
        {
            Scale = Percent,
            MultiplierBase = -Percent,
        },
        new("DamageProtection", ModuleStat.ModuleProtection) { Scale = Percent },
        new("ModuleDefenceAbsorption", ModuleStat.ModuleProtection) { Scale = Percent },

        // Weapons.
        new("Damage", ModuleStat.Damage),
        new("Rounds", ModuleStat.RoundsPerShot) { DefaultBase = 1 },
        new("RoundsPerShot", ModuleStat.RoundsPerShot) { DefaultBase = 1 },
        new("RateOfFire", ModuleStat.RateOfFire),
        new("BurstInterval", ModuleStat.BurstInterval),
        new("BurstSize", ModuleStat.BurstRounds) { DefaultBase = 1 },
        new("BurstRateOfFire", ModuleStat.BurstRateOfFire) { DefaultBase = 1 },
        new("AmmoClipSize", ModuleStat.ClipSize),
        new("AmmoMaximum", ModuleStat.AmmoMaximum),
        new("ReloadTime", ModuleStat.ReloadTime),
        new("DistributorDraw", ModuleStat.DistributorDraw),
        new("ThermalLoad", ModuleStat.ThermalLoad),
        new("ArmourPenetration", ModuleStat.ArmourPiercing),

        // Range is a weapon's maximum range and the journal's alternate spelling of a
        // scanner's ScannerRange. A record resolves it to the stat that family carries.
        new("Range", ModuleStat.MaximumRange),
        new("Range", ModuleStat.ScannerRange),
        new("MaximumRange", ModuleStat.MaximumRange),
        new("FalloffRange", ModuleStat.FalloffRange),

        // A journal spells the same stat DamageFalloffRange, where a blueprint recipe says
        // FalloffRange. Both resolve, as ProbeRadius and DSS_PatchRadius do.
        new("DamageFalloffRange", ModuleStat.FalloffRange),
        new("ShotSpeed", ModuleStat.ShotSpeed),

        // Damage-type shares are percentages in a journal and fractions in the catalogue. They
        // live one level below the module record, so a share reader supplies the value the
        // otherwise flat label mapping reads.
        new("$Kinetic;", null)
        {
            Share = distribution => distribution.Kinetic,
            Scale = Percent,
            DefaultBase = 0,
        },
        new("$Thermal;", null)
        {
            Share = distribution => distribution.Thermal,
            Scale = Percent,
            DefaultBase = 0,
        },
        new("$Explosive;", null)
        {
            Share = distribution => distribution.Explosive,
            Scale = Percent,
            DefaultBase = 0,
        },
        new("$Absolute;", null)
        {
            Share = distribution => distribution.Absolute,
            Scale = Percent,
            DefaultBase = 0,
        },

        // A weapon that carries no jitter fires true, and Rapid Fire, its multi-cannon
        // spelling and Inertial Impact all give one, which a journal confirms by reporting an
        // original value of zero for a missile rack whose record holds no such field. An
        // additive leg starts from zero on its own, so the default is what makes the base
        // explicit: a computed modifier then states the original value the game does.
        new("Jitter", ModuleStat.Jitter) { DefaultBase = 0 },
    ];

    /// <summary>
    /// Every entry for a label, in declaration order. Usually one; <c>Range</c> and
    /// <c>ShieldBankHeat</c> have two each, because the modules carrying those stats keep them
    /// in different fields. The first entry answers everything that does not depend on which
    /// module is asked about.
    /// </summary>
    private static readonly Dictionary<string, List<StatLabel>> ByLabel = BuildIndex();

    /// <summary>
    /// Reads a module's stats as base values keyed by journal modifier label, in the journal's
    /// own units.
    /// </summary>
    /// <param name="module">The module to read.</param>
    /// <returns>One entry per label the module carries a base value for.</returns>
    internal static Dictionary<string, double> BaseStats(OutfittingModule module)
    {
        if (module is null) throw new ArgumentNullException(nameof(module));

        // Only a weapon assumes a value for a stat it leaves out.
        bool weapon = module.Stats.Has(ModuleStat.Damage);
        Dictionary<string, double> values = new(StringComparer.Ordinal);
        foreach (StatLabel entry in Labels)
        {
            if (values.ContainsKey(entry.Label)) continue;

            double? value = entry.Share is null
                ? entry.Stat is null ? null : module.Stats[entry.Stat.Value]
                : module.DamageDistribution is null ? null : entry.Share(module.DamageDistribution);
            if (value.HasValue) values[entry.Label] = value.Value * entry.Scale;
            else if (entry.DefaultBase.HasValue && weapon) values[entry.Label] = entry.DefaultBase.Value;
        }

        return values;
    }

    /// <summary>The module stat a journal modifier label writes back to.</summary>
    /// <param name="label">The journal modifier label.</param>
    /// <param name="stats">
    /// The stats the label is resolved against, when there are any. A label that maps to two
    /// stats, <c>Range</c> or <c>ShieldBankHeat</c>, answers with whichever of them the module
    /// carries; without a module it answers with the first.
    /// </param>
    /// <returns>
    /// The stat, or <see langword="null"/> when the label names a damage share, grants a
    /// capability, or is unknown.
    /// </returns>
    internal static ModuleStat? StatFor(string label, ModuleStats? stats)
    {
        if (!ByLabel.TryGetValue(label, out List<StatLabel>? entries)) return null;
        if (stats is not null && entries.Count > 1)
        {
            foreach (StatLabel entry in entries)
            {
                if (entry.Stat is not null && stats.Has(entry.Stat.Value)) return entry.Stat;
            }
        }

        return entries[0].Stat;
    }

    /// <summary>The journal value divided by the catalogue value for one label.</summary>
    /// <param name="label">The journal modifier label.</param>
    /// <returns>
    /// One hundred for the stats a journal reports as a percentage while the catalogue stores a
    /// fraction, and one for every other label, an unknown one included.
    /// </returns>
    internal static double ScaleFor(string label) => First(label)?.Scale ?? 1;

    /// <summary>The multiplier base a label compounds through.</summary>
    /// <param name="label">The journal modifier label.</param>
    /// <returns>
    /// The base the game divides by before compounding, or <see langword="null"/> for the
    /// ordinary stats that scale directly.
    /// </returns>
    /// <remarks>
    /// These stats do not scale like an ordinary number. A bulkhead's eighty percent hull boost
    /// engineered by a thirty-two percent blueprint becomes 137.6 percent, not 105.6, because
    /// the multiplier 1.8 is what gets multiplied by 1.32. A resistance works the same way on
    /// its damage multiplier, so its base is negative.
    /// </remarks>
    internal static double? MultiplierBaseFor(string label) => First(label)?.MultiplierBase;

    /// <summary>The string a capability-granting label writes to a journal modifier.</summary>
    /// <param name="label">The journal modifier label.</param>
    /// <returns>The written string, or <see langword="null"/> for a numeric stat.</returns>
    /// <remarks>
    /// Anti-Guardian Zone Resistance is a flag with no unit and no magnitude, so its displayed
    /// percentage is not treated as arithmetic.
    /// </remarks>
    internal static string? CapabilityValueFor(string label) => First(label)?.CapabilityValue;

    private static StatLabel? First(string label) =>
        ByLabel.TryGetValue(label, out List<StatLabel>? entries) ? entries[0] : null;

    private static Dictionary<string, List<StatLabel>> BuildIndex()
    {
        Dictionary<string, List<StatLabel>> index = new(StringComparer.Ordinal);
        foreach (StatLabel entry in Labels)
        {
            if (!index.TryGetValue(entry.Label, out List<StatLabel>? entries))
            {
                entries = [];
                index[entry.Label] = entries;
            }

            entries.Add(entry);
        }

        return index;
    }

    /// <summary>One journal modifier label and how it relates to the stat behind it.</summary>
    /// <param name="Label">
    /// The modifier label, such as <c>FSDOptimalMass</c>. These are a journal's own spellings,
    /// except for <c>BurstInterval</c>, which blueprint recipes use for the stat a journal
    /// reports as the resulting rate of fire.
    /// </param>
    /// <param name="Stat">
    /// The module stat holding the base value, or <see langword="null"/> where the label names
    /// a damage share or grants a capability instead.
    /// </param>
    private sealed record StatLabel(string Label, ModuleStat? Stat)
    {
        /// <summary>Reads the damage share the label names, for the labels that name one.</summary>
        internal Func<DamageDistribution, double?>? Share { get; init; }

        /// <summary>The journal value divided by the catalogue value.</summary>
        internal double Scale { get; init; } = 1;

        /// <summary>The base a percentage of a multiplier compounds through.</summary>
        internal double? MultiplierBase { get; init; }

        /// <summary>The value the game assumes when a module carries no such stat.</summary>
        /// <remarks>
        /// A weapon with no rounds per shot fires one round, and one with no burst fires a
        /// single shot per burst. A recipe can modify a defaulted stat, so the default belongs
        /// in the base values a recipe folds.
        /// </remarks>
        internal double? DefaultBase { get; init; }

        /// <summary>The string written to a non-numeric modifier that grants a capability.</summary>
        internal string? CapabilityValue { get; init; }
    }
}
